// apollo-angular-boost/esm5/ApolloBoost.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { ApolloLink, Observable } from 'apollo-link';
import { Apollo } from 'apollo-angular';
import { HttpLink } from 'apollo-angular-link-http';
import { withClientState } from 'apollo-link-state';
import { onError } from 'apollo-link-error';
import { InMemoryCache } from 'apollo-cache-inmemory';
var ApolloBoost = /** @class */ (function () {
    function ApolloBoost(apollo, httpLink) {
        this.apollo = apollo;
        this.httpLink = httpLink;
    }
    /**
     * @param {?} config
     * @return {?}
     */
    ApolloBoost.prototype.create = /**
     * @param {?} config
     * @return {?}
     */
    function (config) {
        /** @type {?} */
        var cache = config && config.cacheRedirects
            ? new InMemoryCache({ cacheRedirects: config.cacheRedirects })
            : new InMemoryCache();
        /** @type {?} */
        var stateLink = config && config.clientState
            ? withClientState(tslib_1.__assign({}, config.clientState, { cache: cache }))
            : false;
        /** @type {?} */
        var errorLink = config && config.onError
            ? onError(config.onError)
            : onError(function (_a) {
                var graphQLErrors = _a.graphQLErrors, networkError = _a.networkError;
                if (graphQLErrors) {
                    graphQLErrors.map(function (_a) {
                        var message = _a.message, locations = _a.locations, path = _a.path;
                        // tslint:disable-next-line
                        return console.log("[GraphQL error]: Message: " + message + ", Location: " + locations + ", Path: " + path);
                    });
                }
                if (networkError) {
                    // tslint:disable-next-line
                    console.log("[Network error]: " + networkError);
                }
            });
        /** @type {?} */
        var requestHandler = config && config.request
            ? new ApolloLink(function (operation, forward) {
                return new Observable(function (observer) {
                    /** @type {?} */
                    var handle;
                    Promise.resolve(operation)
                        .then(function (oper) { return config.request(oper); })
                        .then(function () {
                        handle = forward(operation).subscribe({
                            next: observer.next.bind(observer),
                            error: observer.error.bind(observer),
                            complete: observer.complete.bind(observer),
                        });
                    })
                        .catch(observer.error.bind(observer));
                    return function () {
                        if (handle) {
                            handle.unsubscribe();
                        }
                    };
                });
            })
            : false;
        /** @type {?} */
        var httpLink = this.httpLink.create(tslib_1.__assign({}, ((config && config.httpOptions) || {}), { uri: (config && config.uri) || '/graphql' }));
        /** @type {?} */
        var link = ApolloLink.from((/** @type {?} */ ([
            errorLink,
            requestHandler,
            stateLink,
            httpLink,
        ].filter(function (x) { return !!x; }))));
        this.apollo.create({
            link: link,
            cache: cache,
        });
    };
    ApolloBoost.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    ApolloBoost.ctorParameters = function () { return [
        { type: Apollo },
        { type: HttpLink }
    ]; };
    return ApolloBoost;
}());
export { ApolloBoost };
if (false) {
    /** @type {?} */
    ApolloBoost.prototype.apollo;
    /** @type {?} */
    ApolloBoost.prototype.httpLink;
}
