// apollo-angular-boost/fesm2015/ng.apolloBoost.js
import { InjectionToken, Injectable, NgModule, Optional, Inject } from '@angular/core';
import { ApolloLink, Observable } from 'apollo-link';
export { createOperation, makePromise, toPromise, fromPromise, fromError, Observable, execute, empty, from, split, concat, ApolloLink } from 'apollo-link';
import { Apollo, ApolloModule } from 'apollo-angular';
export { ɵb, ɵa, Apollo, ApolloBase, QueryRef, Query, Mutation, Subscription, SelectPipe, ApolloModule, APOLLO_OPTIONS } from 'apollo-angular';
import { HttpLink, HttpLinkModule } from 'apollo-angular-link-http';
import { withClientState } from 'apollo-link-state';
import { onError } from 'apollo-link-error';
import { InMemoryCache } from 'apollo-cache-inmemory';
export { InMemoryCache, defaultDataIdFromObject, assertIdValue, StoreReader, enhanceErrorWithDocument, WriteError, StoreWriter, HeuristicFragmentMatcher, IntrospectionFragmentMatcher, defaultNormalizedCacheFactory, ObjectCache, record, RecordingCache } from 'apollo-cache-inmemory';
export { default as gql } from 'graphql-tag';
export { printAST, ObservableQuery, NetworkStatus, ApolloError, ApolloClient, FetchType } from 'apollo-client';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/** @type {?} */
const APOLLO_BOOST_CONFIG = new InjectionToken('[apollo-angular-boost] config');

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
class ApolloBoost {
    /**
     * @param {?} apollo
     * @param {?} httpLink
     */
    constructor(apollo, httpLink) {
        this.apollo = apollo;
        this.httpLink = httpLink;
    }
    /**
     * @param {?} config
     * @return {?}
     */
    create(config) {
        /** @type {?} */
        const cache = config && config.cacheRedirects
            ? new InMemoryCache({ cacheRedirects: config.cacheRedirects })
            : new InMemoryCache();
        /** @type {?} */
        const stateLink = config && config.clientState
            ? withClientState(Object.assign({}, config.clientState, { cache }))
            : false;
        /** @type {?} */
        const errorLink = config && config.onError
            ? onError(config.onError)
            : onError(({ graphQLErrors, networkError }) => {
                if (graphQLErrors) {
                    graphQLErrors.map(({ message, locations, path }) => 
                    // tslint:disable-next-line
                    console.log(`[GraphQL error]: Message: ${message}, Location: ${locations}, Path: ${path}`));
                }
                if (networkError) {
                    // tslint:disable-next-line
                    console.log(`[Network error]: ${networkError}`);
                }
            });
        /** @type {?} */
        const requestHandler = config && config.request
            ? new ApolloLink((operation, forward) => new Observable(observer => {
                /** @type {?} */
                let handle;
                Promise.resolve(operation)
                    .then(oper => config.request(oper))
                    .then(() => {
                    handle = forward(operation).subscribe({
                        next: observer.next.bind(observer),
                        error: observer.error.bind(observer),
                        complete: observer.complete.bind(observer),
                    });
                })
                    .catch(observer.error.bind(observer));
                return () => {
                    if (handle) {
                        handle.unsubscribe();
                    }
                };
            }))
            : false;
        /** @type {?} */
        const httpLink = this.httpLink.create(Object.assign({}, ((config && config.httpOptions) || {}), { uri: (config && config.uri) || '/graphql' }));
        /** @type {?} */
        const link = ApolloLink.from((/** @type {?} */ ([
            errorLink,
            requestHandler,
            stateLink,
            httpLink,
        ].filter(x => !!x))));
        this.apollo.create({
            link,
            cache,
        });
    }
}
ApolloBoost.decorators = [
    { type: Injectable }
];
/** @nocollapse */
ApolloBoost.ctorParameters = () => [
    { type: Apollo },
    { type: HttpLink }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
class ApolloBoostModule {
    /**
     * @param {?} boost
     * @param {?=} config
     */
    constructor(boost, config) {
        if (config) {
            boost.create(config);
        }
    }
}
ApolloBoostModule.decorators = [
    { type: NgModule, args: [{
                imports: [ApolloModule, HttpLinkModule],
                providers: [ApolloBoost],
            },] }
];
/** @nocollapse */
ApolloBoostModule.ctorParameters = () => [
    { type: ApolloBoost },
    { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [APOLLO_BOOST_CONFIG,] }] }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */

export { APOLLO_BOOST_CONFIG, ApolloBoost, ApolloBoostModule };
