// apollo-angular-boost/esm2015/ApolloBoost.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { ApolloLink, Observable } from 'apollo-link';
import { Apollo } from 'apollo-angular';
import { HttpLink } from 'apollo-angular-link-http';
import { withClientState } from 'apollo-link-state';
import { onError } from 'apollo-link-error';
import { InMemoryCache } from 'apollo-cache-inmemory';
export class ApolloBoost {
    /**
     * @param {?} apollo
     * @param {?} httpLink
     */
    constructor(apollo, httpLink) {
        this.apollo = apollo;
        this.httpLink = httpLink;
    }
    /**
     * @param {?} config
     * @return {?}
     */
    create(config) {
        /** @type {?} */
        const cache = config && config.cacheRedirects
            ? new InMemoryCache({ cacheRedirects: config.cacheRedirects })
            : new InMemoryCache();
        /** @type {?} */
        const stateLink = config && config.clientState
            ? withClientState(Object.assign({}, config.clientState, { cache }))
            : false;
        /** @type {?} */
        const errorLink = config && config.onError
            ? onError(config.onError)
            : onError(({ graphQLErrors, networkError }) => {
                if (graphQLErrors) {
                    graphQLErrors.map(({ message, locations, path }) => 
                    // tslint:disable-next-line
                    console.log(`[GraphQL error]: Message: ${message}, Location: ${locations}, Path: ${path}`));
                }
                if (networkError) {
                    // tslint:disable-next-line
                    console.log(`[Network error]: ${networkError}`);
                }
            });
        /** @type {?} */
        const requestHandler = config && config.request
            ? new ApolloLink((operation, forward) => new Observable(observer => {
                /** @type {?} */
                let handle;
                Promise.resolve(operation)
                    .then(oper => config.request(oper))
                    .then(() => {
                    handle = forward(operation).subscribe({
                        next: observer.next.bind(observer),
                        error: observer.error.bind(observer),
                        complete: observer.complete.bind(observer),
                    });
                })
                    .catch(observer.error.bind(observer));
                return () => {
                    if (handle) {
                        handle.unsubscribe();
                    }
                };
            }))
            : false;
        /** @type {?} */
        const httpLink = this.httpLink.create(Object.assign({}, ((config && config.httpOptions) || {}), { uri: (config && config.uri) || '/graphql' }));
        /** @type {?} */
        const link = ApolloLink.from((/** @type {?} */ ([
            errorLink,
            requestHandler,
            stateLink,
            httpLink,
        ].filter(x => !!x))));
        this.apollo.create({
            link,
            cache,
        });
    }
}
ApolloBoost.decorators = [
    { type: Injectable }
];
/** @nocollapse */
ApolloBoost.ctorParameters = () => [
    { type: Apollo },
    { type: HttpLink }
];
if (false) {
    /** @type {?} */
    ApolloBoost.prototype.apollo;
    /** @type {?} */
    ApolloBoost.prototype.httpLink;
}
