// apollo-angular-boost/esm2015/index.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import gql from 'graphql-tag';
export { printAST, ObservableQuery, NetworkStatus, ApolloError, ApolloClient, default, FetchType } from 'apollo-client';
export { createOperation, makePromise, toPromise, fromPromise, fromError, Observable, execute, empty, from, split, concat, ApolloLink } from 'apollo-link';
export { defaultDataIdFromObject, assertIdValue, StoreReader, enhanceErrorWithDocument, WriteError, StoreWriter, HeuristicFragmentMatcher, IntrospectionFragmentMatcher, defaultNormalizedCacheFactory, ObjectCache, record, RecordingCache } from 'apollo-cache-inmemory';
export { ɵb, ɵa, Apollo, ApolloBase, QueryRef, Query, Mutation, Subscription, SelectPipe, ApolloModule, APOLLO_OPTIONS } from 'apollo-angular';
export { APOLLO_BOOST_CONFIG } from './tokens';
export { ApolloBoost } from './ApolloBoost';
export { ApolloBoostModule } from './ApolloBoostModule';
export { InMemoryCache } from 'apollo-cache-inmemory';
export { gql };
