// routeshub/esm2015/utils/enhance.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { indexer } from './indexer';
import { checkMultiPath, splitPath } from './path';
import { setState, stateFn } from './state';
/**
 * Serializes routes
 * to enhance capabilities
 * @template R, C
 * @param {?} parentSlice
 * @param {?} routes
 * @return {?}
 */
export function enhance(parentSlice, routes) {
    return Object.keys(routes).reduce((/**
     * @param {?} acc
     * @param {?} routeName
     * @return {?}
     */
    (acc, routeName) => {
        const { children, path, lazy } = routes[routeName];
        /** @type {?} */
        const state = setState(parentSlice, path);
        /** @type {?} */
        const route = {
            id: indexer.next().value,
            parentId: parentSlice !== null ? parentSlice.id : null,
            state,
            stateFn: stateFn.bind(null, state),
            path: checkMultiPath(path) ? splitPath(path) : path,
            lazy: lazy || null,
            routeName
        };
        /* tslint:disable:prefer-object-spread */
        return Object.assign(acc, {
            [routeName]: Object.assign({}, route, { children: children !== undefined ? enhance(route, children) : null })
        });
    }), (/** @type {?} */ ({})));
}
