// routeshub/esm5/utils/enhance.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { indexer } from './indexer';
import { checkMultiPath, splitPath } from './path';
import { setState, stateFn } from './state';
/**
 * Serializes routes
 * to enhance capabilities
 * @template R, C
 * @param {?} parentSlice
 * @param {?} routes
 * @return {?}
 */
export function enhance(parentSlice, routes) {
    return Object.keys(routes).reduce((/**
     * @param {?} acc
     * @param {?} routeName
     * @return {?}
     */
    function (acc, routeName) {
        var _a;
        var _b = routes[routeName], children = _b.children, path = _b.path, lazy = _b.lazy;
        /** @type {?} */
        var state = setState(parentSlice, path);
        /** @type {?} */
        var route = {
            id: indexer.next().value,
            parentId: parentSlice !== null ? parentSlice.id : null,
            state: state,
            stateFn: stateFn.bind(null, state),
            path: checkMultiPath(path) ? splitPath(path) : path,
            lazy: lazy || null,
            routeName: routeName
        };
        /* tslint:disable:prefer-object-spread */
        return Object.assign(acc, (_a = {},
            _a[routeName] = tslib_1.__assign({}, route, { children: children !== undefined ? enhance(route, children) : null }),
            _a));
    }), (/** @type {?} */ ({})));
}
